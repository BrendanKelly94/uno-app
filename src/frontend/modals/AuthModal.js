import React, { useState, useEffect } from "react";
import Modal from "@material-ui/core/Modal";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import ApiEndpoint from "../utils/ApiEndpoint.js";
import CustomError from "../utils/CustomError.js";
import useScale from "../utils/useScale.js";
import useError from "../utils/useError.js";

function AuthModal({ open, setOpen, login }) {
  const [auth, setAuth] = useState({ name: "", pwd: "", rePwd: "" });
  const [reg, setReg] = useState(false);
  const [success, setSuccess] = useState(false);
  const scaleFactor = useScale();
  const [error, errorHandler] = useError();
  const [width, setWidth] = useState(scaleFactor.size < 1.2 ? "90%" : "50%");

  const containerStyle = {
    position: "absolute",
    width: width,
    left: "50%",
    top: "50%",
    transform: "translate(-50%, -50%)"
  };

  const modalStyle = {
    width: "100%",
    display: "flex",
    justifyContent: "center",
    flexDirection: "column",
    backgroundColor: "#fff"
  };

  const errStyle = {
    backgroundColor: "red",
    color: "#fff",
    transform: "scaleY(1)"
  };

  const successStyle = {
    backgroundColor: "green",
    color: "#fff",
    transform: "scaleY(1)"
  };

  const messageStyle = {
    width: "100%",
    height: "2em",
    transform: "scaleY(0)",
    transition: "transform .5s",
    transformOrigin: "bottom",
    textAlign: "center"
  };

  async function handleLogin(){
    errorHandler(async () => {
      const loginData = await new ApiEndpoint("/login").postReq({
        name: auth.name,
        pwd: auth.pwd
      });
      if (loginData.hasOwnProperty("name")) {
        setSuccess(true);
        setTimeout(() => {
          setOpen(false);
        }, 500);
        login(auth.name, auth.pwd);
      }
    })();
  }

  async function handleRegister(){
    errorHandler(async () => {
      if(auth.pwd !== auth.rePwd){
        throw new CustomError({
          status:200,
          message: "Passwords do not match"
        })
      }

      const regData = await new ApiEndpoint("/register").postReq({
        name: auth.name,
        pwd: auth.pwd
      });
      if (regData.hasOwnProperty("name")) {
        setStatus({ err: null, success: true });
        setTimeout(() => {
          setOpen(false);
        }, 500);
        login(auth.name, auth.pwd);
      }
    })();
  }

  useEffect(
    () => {
      if (scaleFactor.size < 1.2) {
        setWidth("90%");
      } else {
        setWidth("50%");
      }
    },
    [scaleFactor]
  );

  return (
    <Modal
      aria-labelledby="log into account"
      aria-describedby="log in or register popup"
      open={open}
      onClose={setOpen}
    >
      <div style={containerStyle}>
        {
          <div
            style={
              error.status
                ? { ...messageStyle, ...errStyle }
                : success
                ? { ...messageStyle, ...successStyle }
                : messageStyle
            }
          >
            {error.status && !success && <div>{error.message}</div>}
            {success && <div>{"You have successfully logged in!"}</div>}
          </div>


        }
        <div style={modalStyle}>
          <TextField
            id="user-name-input"
            label="UserName"
            margin="normal"
            variant="outlined"
            value={auth.name}
            style={{ margin: ".5em" }}
            onChange={e => setAuth({ ...auth, name: e.target.value })}
          />
          <TextField
            id="password-input"
            label="Password"
            type="password"
            margin="normal"
            variant="outlined"
            value={auth.pwd}
            style={{ margin: ".5em" }}
            onChange={e => setAuth({ ...auth, pwd: e.target.value })}
          />
          {reg && (
            <TextField
              id="re-password-input"
              label="Re-Password"
              type="password"
              margin="normal"
              variant="outlined"
              value={auth.rePwd}
              style={{ margin: ".5em" }}
              onChange={e => setAuth({ ...auth, rePwd: e.target.value })}
            />
          )}
          <Button onClick={reg ? handleRegister : handleLogin}>
            {reg ? "Create Account" : "Log In"}
          </Button>
          <button onClick={() => setReg(!reg)}>
            {reg ? "Already have account?" : "Don't have account?"}
          </button>
        </div>
      </div>
    </Modal>
  );
}

export default AuthModal;
