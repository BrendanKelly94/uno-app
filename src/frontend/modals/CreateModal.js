import React, { useState } from 'react';
import Modal from '@material-ui/core/Modal';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Checkbox from '@material-ui/core/Checkbox';
import Button from '@material-ui/core/Button';
import history from '.././utils/history.js';

function CreateModal({ open, setOpen }){
  const [ fillBots, setFillBots ] = useState(false);
  const [ err, setErr ] = useState(null);

  const containerStyle = {
    position: 'absolute',
    width: '50%',
    left: '50%',
    top: '50%',
    transform: 'translate(-50%, -50%)'
  }

  const modalStyle = {
    width: '100%',
    display: 'flex',
    justifyContent: 'center',
    flexDirection: 'column',
    backgroundColor: '#fff'
  }


  const handleCreate = async () => {
    const res = await fetch('http://localhost:3000/api/newGame', {
        method: "POST",
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({botFill: fillBots})
    })
    const json = await res.json();
    if(json.hasOwnProperty(err)){
      setError(json.err)
    }else{
      history.push(`game/${json.id}`);
    }
  }

  return(
    <Modal
          aria-labelledby="log into account"
          aria-describedby="log in or register popup"
          open={open}
          onClose={setOpen}
        >
        <div style={containerStyle}>
          <div style={modalStyle} >
            <button onClick = {() => setOpen()}> x </button>
            <FormControlLabel
              control={
                <Checkbox
                  checked={fillBots}
                  onChange={() => setFillBots(!fillBots)}
                  value="checkedB"
                  color="primary"
                />
                }
                label="fill with bots"
            />
            <Button onClick = {handleCreate}>Create Game</Button>
          </div>
        </div>
    </Modal>
  )
}

export default CreateModal;
