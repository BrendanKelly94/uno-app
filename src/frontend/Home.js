import React, { useState, useContext } from 'react';
import AuthModal from './modals/AuthModal';
import CreateModal from './modals/CreateModal';
import authStoreContext from './context/authStoreContext';
import history from './utils/history.js';


function Home(){
  const [auth, setAuth] = useState(false);
  const [create, setCreate] = useState(false);
  const [context, setContext] = useContext(authStoreContext);

  const handleJoin = () => {
    if(context.user_name){
      history.push('/gamesList')
    }else{
      setAuth(!auth);
    }
  }

  const handleCreate = () => {
    if(context.user_name){
      setCreate(!create);
    }else{
      setAuth(!auth);
    }
  }

  const login = (uName, pwd) => {
    setContext({user_name: uName, pwd: pwd});
  }

  return(
    <div>
      <button id = 'join-button' onClick = {handleJoin}>Join Game</button>
      <button id = 'create-button' onClick = {handleCreate}>Create Game</button>
      <AuthModal open = {auth} setOpen = {() => setAuth(!auth)} login = {login}/>
      <CreateModal open = {create} setOpen = {() => setCreate(!create)}/>
    </div>
  )
}

export default Home;
