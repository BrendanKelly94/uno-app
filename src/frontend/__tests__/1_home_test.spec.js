import React from 'react';
import ReactDOM from 'react-dom';
import { configure, shallow, mount, render } from 'enzyme';
import Home from '.././Home';
import AuthModal from '../modals/AuthModal';
import CreateModal from '../modals/CreateModal';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('<Home/>', () => {
  beforeEach(() => {
      window.sessionStorage = {}
  })
  it('will render component', () => {
    const wrapper = render(<Home />);
    expect(wrapper).toMatchSnapshot();
  })
  it('will render join game button', () => {
    const wrapper = render(<Home />)
    expect(wrapper.find('#join-button')).toBeDefined();
  });
  it('will render create game button', () => {
    const wrapper = render(<Home/>)
    expect(wrapper.find('#create-button')).toBeDefined();

  });
  it('will render modal if user clicks button and is not logged in', () => {
    const wrapper = mount(<Home/>);
    wrapper.find('#join-button').simulate('click');
    expect(wrapper.find(AuthModal).props().open).toBeDefined();
  })
  it('will take you to join game route if logged in and click join button', () => {
    window.sessionStorage = {name: 'test'}
    const wrapper = render(<Home/>)
    expect(true).toBeFalsy();
  });
  it('will open create game modal if logged in and click create game button', () => {
    window.sessionStorage = {name: 'test'}
    const wrapper = mount(<Home/>);
    expect(wrapper.find(CreateModal).props().open).toBeDefined();
  })
})
