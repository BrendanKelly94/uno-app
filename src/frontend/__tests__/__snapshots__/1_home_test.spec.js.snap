// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home/> will render component 1`] = `
<div>
  <button
    id="join-button"
  >
    Join Game
  </button>
  <button
    id="create-button"
  >
    Create Game
  </button>
  <div />
</div>
`;
